tests/activation/test_prompts.py

```diff
@@ -0,0 +1,280 @@
+"""test that prompt behavior is correct in supported shells"""
+from __future__ import absolute_import, unicode_literals
+
+import os
+import subprocess
+import sys
+from collections import namedtuple
+from textwrap import dedent
+
+import pytest
+
+import virtualenv
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+
+# This must match the DEST_DIR provided in the ../conftest.py:clean_python fixture
+ENV_DEFAULT = "env"
+
+# This can be anything
+ENV_CUSTOM = "env_custom"
+
+# Standard prefix, surround the env name in parentheses and separate by a space
+PREFIX_DEFAULT = "({}) ".format(ENV_DEFAULT)
+
+# Arbitrary prefix for the environment that's provided a 'prompt' arg
+PREFIX_CUSTOM = "---ENV---"
+
+# Filename template: {shell}.script.(normal|suppress).(default|custom)[extension]
+SCRIPT_TEMPLATE = "{}.script.{}.{}{}"
+
+# Filename template: {shell}.out.(normal|suppress).(default|custom)
+OUTPUT_TEMPLATE = "{}.out.{}.{}"
+
+SHELL_LIST = ["bash", "fish", "csh", "xonsh", "cmd", "powershell"]
+
+
+# Py2 doesn't like unicode in the environment
+def env_compat(s):
+    return s.encode("utf-8") if sys.version_info.major < 3 else s
+
+
+@pytest.fixture(scope="module")
+def platform_check_skip(tmp_path_factory):
+    """Return function triggering skip based on platform & shell."""
+    platform_incompat = "No sane provision for {} on {} yet"
+
+    def check(platform, shell):
+
+        if (sys.platform.startswith("win") and shell in ["bash", "csh", "fish"]) or (
+            sys.platform.startswith("linux") and shell in ["cmd", "powershell"]
+        ):
+            pytest.skip(platform_incompat.format(shell, platform))
+
+        if sys.platform.startswith("win") and shell == "xonsh":
+            pytest.skip("Provisioning xonsh on windows is unreliable")
+
+        if shell == "xonsh" and sys.version_info < (3, 4):
+            pytest.skip("xonsh requires Python 3.4 at least")
+
+        if shell == "powershell":
```

Comment:
this feels like hard to maintain, branches for various shells, can we somehow group testing logic together for per shell 🤔 to avoid the need for ifs that is 👍

Reply:
Not sure what you mean -- this fixture is just to handle the logic of skipping certain test combinations based on platform and shell.  I think it's about as centralized as it can get, and it's not like any of the other shells care about, e.g., whether powershell has scripting permissions enabled.

It could be split up, I guess, where there's a separate function for each shell, defining its own skip conditions... folded into the `shell_info` namedtuple as `.platform_skip`, maybe?  And then the tests could call `shell_info[shell].platform_skip()`?

Reply:
Is [this](https://github.com/bskinn/virtualenv/blob/5e5b5c17da7674f7d1e55d835baca48f2b6d3dad/tests/activation/test_prompts.py#L67-L97) more like what you had in mind, @gaborbernat?

Else, I could add a new field to [`ShellInfo`](https://github.com/bskinn/virtualenv/blob/5e5b5c17da7674f7d1e55d835baca48f2b6d3dad/tests/activation/test_prompts.py#L138-L148), say `skip_check`, and populate with functions encapsulating the skip logic per-shell.

Reply:
Restructured implementation in 5e5b5c1